lib/common/httpclient.c

```diff
@@ -0,0 +1,199 @@
+ * IN THE SOFTWARE.
+ */
+
+#include "h2o/httpclient_internal.h"
+#include "h2o/httpclient_internal_h1.h"
+#include "h2o/httpclient_internal_h2.h"
+
+const char *const h2o_httpclient_error_is_eos = "end of stream";
+const char *const h2o_httpclient_error_refused_stream = "refused stream";
+
+void h2o_httpclient_connection_pool_init(h2o_httpclient_connection_pool_t *connpool, h2o_socketpool_t *sockpool)
+{
+    connpool->socketpool = sockpool;
+    h2o_linklist_init_anchor(&connpool->http2.conns);
+}
+
+static void close_client(struct st_h2o_httpclient_private_t *client)
+{
+    if (client->connect_req != NULL) {
+        h2o_socketpool_cancel_connect(client->connect_req);
+        client->connect_req = NULL;
+    }
+
+    if (h2o_timeout_is_linked(&client->connect_timeout_entry))
+        h2o_timeout_unlink(&client->connect_timeout_entry);
+
+    free(client);
+}
+
+static void on_connect_error(struct st_h2o_httpclient_private_t *client, const char *errstr)
+{
+    assert(errstr != NULL);
+    client->cb.on_connect(&client->super, errstr, NULL, NULL, NULL, 0, NULL, NULL, NULL, NULL);
+    close_client(client);
+}
+
+static void on_connect_timeout(h2o_timeout_entry_t *entry)
+{
+    struct st_h2o_httpclient_private_t *client = H2O_STRUCT_FROM_MEMBER(struct st_h2o_httpclient_private_t, connect_timeout_entry, entry);
+    on_connect_error(client, "connection timeout");
+}
+
+static void do_cancel(h2o_httpclient_t *_client)
+{
+    struct st_h2o_httpclient_private_t *client = (void *)_client;
+    close_client(client);
+}
+
+static struct st_h2o_httpclient_private_t *create_client(h2o_mem_pool_t *pool, void *data, h2o_httpclient_ctx_t *ctx, h2o_httpclient_connect_cb cb)
+{
+#define SZ_MAX(x, y) ((x) > (y) ? (x) : (y))
+    size_t sz = SZ_MAX(sizeof(struct st_h2o_http1client_t), sizeof(struct st_h2o_http2client_stream_t));
+#undef SZ_MAX
+    struct st_h2o_httpclient_private_t *client = h2o_mem_alloc(sz);
+    memset(client, 0, sizeof(*client));
+    client->super.pool = pool;
+    client->super.ctx = ctx;
+    client->super.data = data;
+    client->super.cancel = do_cancel;
+    client->super.steal_socket = NULL;
+    client->super.update_window = NULL;
+    client->super.write_req = NULL;
+    client->cb.on_connect = cb;
+    client->connect_timeout_entry.cb = on_connect_timeout;
+
+    return client;
+}
+
+static void on_pool_connect(h2o_socket_t *sock, const char *errstr, void *data, h2o_url_t *origin)
+{
+    struct st_h2o_httpclient_private_t *client = data;
+
+    h2o_timeout_unlink(&client->connect_timeout_entry);
+
+    client->connect_req = NULL;
+
+    if (sock == NULL) {
+        assert(errstr != NULL);
+        on_connect_error(client, errstr);
+        return;
+    }
+
+    h2o_iovec_t alpn_proto;
+    if (sock->ssl == NULL || (alpn_proto = h2o_socket_ssl_get_selected_protocol(sock)).len == 0) {
+        h2o_http1client_on_connect(client, sock, origin);
+    } else {
+        if (memcmp(alpn_proto.base, "h2", alpn_proto.len) == 0) {
```

Comment:
Please use `h2o_memis`, unless we can be very sure that `alpn_proto.len` is longer than two octets. Do we have such an assumption?

Reply:
fixed at https://github.com/h2o/h2o/pull/1549/commits/e3b5ab0dc8e3b3b2b0af4427742a7cd21c59e37a